sdk/identity/Azure.Identity/src/TokenCache.cs

```diff
@@ -0,0 +1,289 @@
+
+        /// <summary>
+        /// Deserializes the <see cref="TokenCache"/> from the specified <see cref="Stream"/>.
+        /// </summary>
+        /// <param name="stream">The <see cref="Stream"/> from which the serialized <see cref="TokenCache"/> will be read.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> controlling the request lifetime.</param>
+        public static async Task<TokenCache> DeserializeAsync(Stream stream, CancellationToken cancellationToken = default)
+        {
+            if (stream is null)
+                throw new ArgumentNullException(nameof(stream));
+
+            return await DeserializeAsync(stream, true, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task SerializeAsync(Stream stream, bool async, CancellationToken cancellationToken)
+        {
+            if (async)
+            {
+                await stream.WriteAsync(_data, 0, _data.Length, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                stream.Write(_data, 0, _data.Length);
+            }
+        }
+
+        private static async Task<TokenCache> DeserializeAsync(Stream stream, bool async, CancellationToken cancellationToken)
+        {
+            var data = new byte[stream.Length - stream.Position];
+
+            if (async)
+            {
+                await stream.ReadAsync(data, 0, data.Length, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                stream.Read(data, 0, data.Length);
+            }
+
+            return new TokenCache(data);
+        }
+
+        internal virtual async Task RegisterCache(bool async, ITokenCache tokenCache, CancellationToken cancellationToken)
+        {
+            if (async)
+            {
+                await _lock.WaitAsync(cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                _lock.Wait(cancellationToken);
+            }
+
+            try
+            {
+                if (!_cacheAccessMap.TryGetValue(tokenCache, out _))
+                {
+                    tokenCache.SetBeforeAccessAsync(OnBeforeCacheAccessAsync);
+
+                    tokenCache.SetAfterAccessAsync(OnAfterCacheAccessAsync);
+
+                    _cacheAccessMap.Add(tokenCache, new CacheTimestamp());
+                }
+            }
+            finally
+            {
+                _lock.Release();
+            }
+        }
+
+        private async Task OnBeforeCacheAccessAsync(TokenCacheNotificationArgs args)
+        {
+            if (_disposedValue)
+            {
+                throw new ObjectDisposedException(nameof(TokenCache));
+            }
+
+            await _lock.WaitAsync().ConfigureAwait(false);
+
+            try
+            {
+                args.TokenCache.DeserializeMsalV3(_data, true);
+
+                _cacheAccessMap.GetOrCreateValue(args.TokenCache).Update();
```

Comment:
Can we use `AddOrUpdate` here and avoid having the `CacheTimestamp` class?

Reply:
conditional week table doesn't have `AddOrUpdate`

Reply:
https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.conditionalweaktable-2.addorupdate?view=net-5.0 

Ah, it's in NS2.1...

Reply:
Perhaps I should have said intellisense didn't show me `AddOrUpdate` 😉